doc/octopus_setbg.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH octopus_setbg 1 "January 20, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
octopus_setbg \- set the background image of the desktop
.SH SYNOPSIS
.B octopus_setbg
.RB [ options ]
.RB image_file
.SH DESCRIPTION
.B octopus_setbg
is a small utility to set the background image of your desktop for the
.B octopus
window manager. If possible it scales the image so that it fills completely
the desktop. The image file could be in any format which your gdk_pixbuf image loader
supports.
.SH OPTIONS
.PP
.TP 8
.B \-c
Do not center the image.
.TP 8
.B \-s
Do not scale image if aspect ratio of image and screen matches
.TP 8
.B \-h
.TP 8
.B \-?
Show a short help message.
.SH "ENVIRONMENT VARIABLES"
.IP DISPLAY
This variable is used to determine which X server to use.
.SH "SEE ALSO"
.BR octopus (1),
.BR octopus_xkeyd (1).
.SH AUTHOR
octopus_setbg was written by Dr. Johann Pfefferl <pfefferl@gmx.net>.
.PP
This manual page was written by Dr. Johann Pfefferl <pfefferl@gmx.net>,
for the Debian project (but may be used by others).


debian/changelog
octopus (0.4.3-5) unstable; urgency=low

  * Manual pages describe now also the available mouse controls

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sat, 24 Jan 2009 21:47:30 +0100

octopus (0.4.3-4) unstable; urgency=low

  * Bugfix: "click-to-type" focus is lost when client was added to an already
    existing frame
  * Manual pages updated

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sat, 24 Jan 2009 17:18:09 +0100

octopus (0.4.3-3) unstable; urgency=low

  * Added manual pages and example configuration files for the provided programs.
  * Added missing documentation for keyboard shortcut Mod-Escape

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Thu, 22 Jan 2009 19:19:59 +0100

octopus (0.4.3-2) unstable; urgency=low

  * Focus new client if the mouse is pressed inside this client
  * React only any of the 3 buttons

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sun, 18 Jan 2009 16:00:18 +0100

octopus (0.4.3-1) unstable; urgency=low

  * Keyboard shortcuts added for vdesk switching if more than 4 virtual
    desktops are configured
  * vdesk change is also possible with keys 1 to 9 (Not only with the
    function keys F1 to F12. Some keyboards do not possess good function
    keys (for example the smaller netbooks).

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Mon,  5 Jan 2009 12:48:02 +0100

octopus (0.4.2-1) unstable; urgency=low

  * The resize fraction in xinerama mode depends of the actual screen
    size and not of the whole display.

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Wed,  6 Aug 2008 11:37:20 +0200

octopus (0.4.1-4) unstable; urgency=low

  * Segmentation fault fixed in init_xml_config function

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Mon,  5 May 2008 15:51:12 +0200

octopus (0.4.1-3) unstable; urgency=low

  * Correct handling of windows with PMaxSize attribute (for example Skype)
  * Better placement of application with USPosition attribute and window
    type "normal"

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Thu, 24 Apr 2008 14:38:49 +0200

octopus (0.4.1-2) unstable; urgency=low

  * Increased window placement a lot (especially Qt windows)
  * Fullscreen/Maximize handling improved a lot

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Fri,  7 Mar 2008 15:42:39 +0100

octopus (0.4.1-1) unstable; urgency=low

  * Window placement has been reworked

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Mon, 18 Feb 2008 15:57:38 +0100

octopus (0.4.0-2) unstable; urgency=low

  * Small bug fixed on systems with no activ xinerama display

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Thu, 14 Feb 2008 10:32:18 +0100

octopus (0.4.0-1) unstable; urgency=low

  * Added Xinerama support
  * Improved placement of new windows

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Wed, 13 Feb 2008 23:51:32 +0100

octopus (0.3.2-2) unstable; urgency=low

  * The kill-client action has now the shortcut "Alt-Escape" instead of
    "Alt-Delete"

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sun, 15 Jul 2007 18:32:27 +0200

octopus (0.3.2-1) unstable; urgency=low

  * The keybindings were changed
  * Better fullscreen handling

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sun, 15 Jul 2007 09:13:45 +0200 

octopus (0.3.1-1) unstable; urgency=low

  * Handle fullscreen mode correctly for applications with an aspect ratio
  * New "jump" operation added

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sun, 01 Jul 2007 13:24:04 +0200

octopus (0.3.0-2) unstable; urgency=low

  * Tool octopus_xkeyd can handle keycode specifications in the config file

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sat, 03 Feb 2007 14:10:15 +0100

octopus (0.3.0-1) unstable; urgency=low

  * Added new tool octopus_xkeyd to bind actions to special X11 keys

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sat, 20 Jan 2007 20:35:46 +0100

octopus (0.2.2-3) unstable; urgency=low

  * Focus follows correctly now when a window is dragged onto another frame

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Mon, 01 Jan 2007 17:44:23 +0100

octopus (0.2.2-2) unstable; urgency=low

  * Fixed nasty focus jumping bug during desktop switching

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Mon, 01 Jan 2007 16:17:08 +0100

octopus (0.2.2-1) unstable; urgency=low

  * Hopefully solved the problem with two focused clients

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sun, 31 Dec 2006 12:17:56 +0100

octopus (0.2.1-3) unstable; urgency=low

  * Filter away EnterWindowMask events after desktop switching

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sat, 25 Nov 2006 13:32:09 +0100

octopus (0.2.1-2) unstable; urgency=low

  * Corrected lost X events after desktop switching

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Fri, 24 Nov 2006 21:23:25 +0100

octopus (0.2.1-1) unstable; urgency=low

  * Changed focus change problems after virtual desktop switching

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sun, 19 Nov 2006 15:51:30 +0100

octopus (0.2.0-5) unstable; urgency=low

  * Added Debian postinst and prerm scripts

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Wed, 29 Mar 2006 21:23:48 +0200

octopus (0.2.0-4) unstable; urgency=low

  * Added this tool to the x-window-manager group

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Wed, 29 Mar 2006 21:00:05 +0200

octopus (0.2.0-3) unstable; urgency=low

  * Added utility to set the screen background

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Tue, 24 Jan 2006 21:07:50 +0100

octopus (0.2.0-2) unstable; urgency=low

  * Multiple transient windows are nowe handled correctly

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sun, 22 Jan 2006 08:47:40 +0100

octopus (0.2.0-1) unstable; urgency=low

  * octopus now uses the official libcairo version 1.0.0 with an official,
    stable API

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Mon, 02 Jan 2006 14:59:23 +0100

octopus (0.1.0-1) unstable; urgency=low

  * Initial Release.

 -- Dr. Johann Pfefferl <pfefferl@gmx.net>  Sat, 16 Jul 2005 20:24:35 +0200



utils/octopus_setbg.c
/* Compile with:
   gcc -Wall -o octopus-setbg octopus-setbg.c -L/usr/X11R6/lib -lX11  $(pkg-config --cflags --libs cairo glib-2.0)
   */
#define G_LOG_DOMAIN "octopus-setbg"
#include <glib.h>
#include <gdk-pixbuf/gdk-pixbuf.h>
#include <gdk-pixbuf-xlib/gdk-pixbuf-xlibrgb.h>
#include <unistd.h>
#include <stdlib.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
/*#include <cairo-xlib.h>*/

static void set_prop(Display *dpy, Window win, Atom at, Atom type, long val)
{
  XChangeProperty(dpy, win, at, type, 32,
      PropModeReplace, (unsigned char *) &val, 1);
}

static Pixmap read_old_pixmap(Display *dpy, Window rootwin, Atom at)
{
  Atom real_type;
  gint real_format;
  gulong items_read, items_left;
  guint8 *data = NULL;
  Pixmap value = 0;

  if(Success == XGetWindowProperty(dpy, rootwin, at,
        0L, 1L, False, XA_PIXMAP, &real_type, &real_format, &items_read, &items_left,
        &data) && items_read) {
    value = *(Pixmap*)data;
    XFree(data);
  }
  return value;
}

static void show_help(gchar *prgname)
{
  g_printerr("Usage: %s [options] <img_filename>\n", prgname);
  g_printerr("Sets the workspace background to the specified image\n\n");
  g_printerr(" -c don't center image\n");
  g_printerr(" -s don't scale image if aspect ratio of image and screen matches\n");
  g_printerr(" -h show this help\n -? show this help\n");
  exit(4);
}

static void install_new_bg_rootpixmap(Display *dpy, Window rootwin, Pixmap pmap)
{
  Atom x_rootpmap_id = XInternAtom(dpy, "_XROOTPMAP_ID", False);
  Pixmap old_pixmap = read_old_pixmap(dpy, rootwin, x_rootpmap_id);

  XSetWindowBackgroundPixmap(dpy, rootwin, pmap);
  XClearWindow(dpy, rootwin);
  XSync(dpy, False);
  XGrabServer(dpy);
  old_pixmap = read_old_pixmap(dpy, rootwin, x_rootpmap_id);
  g_message("old_pixmap=0x%lx is replaced by 0x%lx", old_pixmap, pmap);
  XKillClient(dpy, old_pixmap);
  XSync(dpy, False);
  set_prop(dpy, rootwin, x_rootpmap_id, XA_PIXMAP, pmap);
  XUngrabServer(dpy);
  XFlush(dpy);
  /* Don't delete the created background pixmap */
  XSetCloseDownMode(dpy, RetainPermanent);
}

int main(gint argc, gchar **argv)
{
  Display *dpy;
  Window rootwin;
  gint o, scr;
  Pixmap pmap;
  gint swidth, sheight, iw, ih;
  gchar *img_fname = NULL;
  GdkPixbuf *imgpb;
  GError *error = NULL;
  gboolean center = TRUE, scale = TRUE;

  g_type_init();
  while(0 <= (o = getopt(argc, argv, "csh?"))) {
    switch(o) {
      case 'c':
        center = FALSE;
        break;
      case 's':
        scale = FALSE;
        break;
      default:
        show_help(argv[0]);
        break;
    }
  }
  if(optind >= argc) /* An image filename must be given */
    show_help(argv[0]);
  else
    img_fname = argv[optind++];
  if(!(dpy = XOpenDisplay(0))) {
    g_critical("XOpenDisplay failed");
    return 1;
  }
  scr = DefaultScreen(dpy);
  rootwin = RootWindow(dpy, scr);
  swidth = DisplayWidth(dpy, scr);
  sheight = DisplayHeight(dpy, scr);
  xlib_rgb_init(dpy, ScreenOfDisplay(dpy, scr));
  g_message("%s: scr=%d swidth=%d sheight=%d", img_fname, scr, swidth, sheight);
  imgpb = gdk_pixbuf_new_from_file(img_fname, &error);
  if(error) {
    g_critical(error->message);
    return 2;
  }
  iw = gdk_pixbuf_get_width(imgpb);
  ih = gdk_pixbuf_get_height(imgpb);
  g_message("has_alpha=%d nr_bits_per_pixel=%d row_stride=%d n_channels=%d w=%d h=%d",
      gdk_pixbuf_get_has_alpha(imgpb),
      gdk_pixbuf_get_bits_per_sample(imgpb),
      gdk_pixbuf_get_rowstride(imgpb),
      gdk_pixbuf_get_n_channels(imgpb),
      iw, ih);
  if(scale && (iw != swidth || ih != sheight) && iw * sheight == swidth * ih) { /* Aspect ratio is the same */
    GdkPixbuf *scaled_img = gdk_pixbuf_scale_simple(imgpb, swidth, sheight, GDK_INTERP_BILINEAR);
    if(scaled_img) {
      imgpb = scaled_img;
      iw = gdk_pixbuf_get_width(imgpb);
      ih = gdk_pixbuf_get_height(imgpb);
      g_message("Scaled: has_alpha=%d nr_bits_per_pixel=%d row_stride=%d n_channels=%d w=%d h=%d",
          gdk_pixbuf_get_has_alpha(imgpb),
          gdk_pixbuf_get_bits_per_sample(imgpb),
          gdk_pixbuf_get_rowstride(imgpb),
          gdk_pixbuf_get_n_channels(imgpb),
          iw, ih);
    }
  }

  pmap = XCreatePixmap(dpy, DefaultRootWindow(dpy), swidth, sheight, DefaultDepth(dpy, scr));
  if(pmap) {
    gboolean has_alpha;
    gint row_stride, nr_channels, xpos, ypos;
    guint8 *img_data = NULL, *pixels;
#if 0
    guint8 *rp, *p;
    gint r, c;
    guint32 *img_data_elem;
#endif

    g_message("Pixmap creation successful: pixmap id = 0x%lx", pmap);
    row_stride = gdk_pixbuf_get_rowstride(imgpb);
    nr_channels = gdk_pixbuf_get_n_channels(imgpb);
    has_alpha = gdk_pixbuf_get_has_alpha(imgpb);
    pixels = gdk_pixbuf_get_pixels(imgpb);
    g_message("has_alpha=%d nr_bits_per_pixel=%d row_stride=%d n_channels=%d",
        has_alpha,
        gdk_pixbuf_get_bits_per_sample(imgpb),
        row_stride, nr_channels);

    XFillRectangle(dpy, pmap, DefaultGC(dpy, scr), 0, 0, swidth, sheight);
    xpos = center && iw < swidth ?  (swidth-iw)/2  : 0;
    ypos = center && ih < sheight ? (sheight-ih)/2 : 0;
    if(G_UNLIKELY(has_alpha))
      xlib_draw_rgb_32_image(pmap, DefaultGC(dpy, scr),
          xpos, ypos, iw, ih,
          XLIB_RGB_DITHER_NONE, pixels, row_stride);
    else
      xlib_draw_rgb_image(pmap, DefaultGC(dpy, scr),
          xpos, ypos, iw, ih,
          XLIB_RGB_DITHER_NONE, pixels, row_stride);
#if 0
    img_data = g_malloc(swidth*sheight*sizeof(guint32));
    rp = pixels;
    for(r = 0; r < ih; r++, rp += row_stride) {
      img_data_elem = (guint32*)img_data + r * swidth;
      for(p = rp, c = 0; c < iw; c++) {
        guint32 alpha;
        guint32 red = *p++;
        guint32 green = *p++;
        guint32 blue = *p++;
        if(G_UNLIKELY(has_alpha))
          alpha = *p++;
        else
          alpha = 0xff;
        *img_data_elem++ = (alpha<<24) | (red<<16) | (green<<8) | blue;
      }
    }
#endif
    g_object_unref(imgpb);
#if 0
    {
      cairo_surface_t *img = cairo_image_surface_create_for_data(
          img_data, CAIRO_FORMAT_ARGB32,
          swidth, sheight, swidth * sizeof(guint32));
      cairo_pattern_t *pat = cairo_pattern_create_for_surface(img);
      cairo_surface_t *cr_xlib_surf = cairo_xlib_surface_create(dpy, pmap,
          DefaultVisual(dpy, scr), swidth, sheight);
      cairo_t *cr = cairo_create(cr_xlib_surf);

      cairo_surface_destroy(cr_xlib_surf); /* A reference is hold by cr */
      cairo_surface_destroy(img);
      /* Black background */
      cairo_set_source_rgba(cr, 0., 0., 0., 1.);
      cairo_rectangle(cr, 0, 0, swidth, sheight);
      cairo_fill(cr);
      /* Draw image onto canvas */
      cairo_rectangle(cr, 0, 0, swidth, sheight);
      cairo_set_source(cr, pat);
      cairo_set_antialias(cr, CAIRO_ANTIALIAS_NONE);
      cairo_fill(cr);
      cairo_pattern_destroy(pat);
      cairo_destroy(cr);
    }
#endif
    install_new_bg_rootpixmap(dpy, rootwin, pmap);
    g_free(img_data);
  }

  XCloseDisplay(dpy);
  return 0;
}
